Retain Rendering World

re0312 opened the pull request:
# Objective

- fix #12144 ,compare to #14252 
- We want Relations! However, the render world currently hinders the development of  `* as entities` because it needs to despawn all entities each frame to ensure consistent entity IDs between the main world and render world.


## Solution
- Transition to a strategy of maintaining entity-to-entity mapping between the main world and render world, abandoning the method of removing all entities every frame.
- Use `Observer `to listen the components that tis entity needs to be synchronized to the render world.
- Before executing the extract schedule, the `entity_sync_system `will be run first, which is responsible for maintaining the entity mapping between the main world and the rendering world

Bevy Schedule 
![image](https://github.com/user-attachments/assets/013e9822-e575-43a5-bb0c-108dfc167fd1)


##  Alternative(s)

we could using Resource to manage all rendering entity component data  so that we don't have any entities at all.

## Advantages

- We can continue to use ECS in rendering, which can bring more performance advantages in many aspects(Retain redering world + update in place)
- More consistent primitives, as Bevy is entirely based on ECS.

## Drawbacks 
- Once a component has a dependency relationship on other main world entities, this relationship must also be synchronized to the rendering world, which undoubtedly increases the user's mental burden
- Not all the entities are in sync with the rendering world,  And that makes it pretty confusing to tell apart the entities in the main world from those in the rendering world when we're in the middle of rendering stuff. And considering our extensive use of  `EntityHashmap `to handle rendering data. that's could made everything even more of a headache to deal with.

## Performance
cargo run --release --example many_cubes --features bevy/trace_tracy
red is main
![image](https://github.com/user-attachments/assets/dafe7d09-3fd3-4245-b651-615409036b8c)

cargo run --release --example many_foxes --features bevy/trace_tracy
![image](https://github.com/user-attachments/assets/da5feaa4-a9ba-4fec-a297-1c3f253169bd)


cargo run --release --example many_buttons --features bevy/trace_tracy 

![image](https://github.com/user-attachments/assets/8072a345-b8d4-4af8-8fed-86d8cff4359a)

~30% loss
The majority of the performance loss from the spawn(~3ms) and despawn(~3ms) operations of ephemeral entity.

## Next Steps

1. Stop spawn temporary entities and depsawn it every frame , it could persist in render world
2. Update in place for instance data instead re-insert it per frame 


## Migration Guide

Render world no longer clear all entities per frame , Instead, You must manually control the spawn and despawn of render entities properly.

Before this pr ,`extract` would copy the related main world entity and its data into render world. Then, the render world clear all render entities at the end of the frame to reserve enough entity space, ensuring that no main world entity ID would be occupied when next `extract`.

For the sake of `* as entities`, **we turn to an entity-to-entity mapping strategy to sync between main world entity and render world entity, where each `synchronized` main entity has a `RenderEntity` component  that holds an Entity ID pointer to its unique counterpart entity in the render world.**

![image](https://github.com/user-attachments/assets/3aac38a3-05b2-4a28-a54a-333669e5887a)

To establish a "Synchronous Relationship", you can add a `SyncRenderWorld` component to an entity, indicating that it needs to be synchronized to the render world. (Currently, its added by default in `Camera2dBundle` , `Camera3dBundle`, `DirectionalLightBundle`, `SpotLightBundle`, `PointLightBundle`, `SpriteBundle`, and `SpriteSheetBundle`.)

 Once a `synchronized` main entity despawned ,its corresponding render entity will automatically be purged before next `extract`.

Temporary entities are part of entities in render world  , bevy's renderer spawns numerous temporary entities for rendering per frame , which need to be cleared at the end of each frame.
To avoid duplicating spawn these entities each frame,you can add a `TemporaryRenderEntity` component to it. Bevy will then automatically clean up it at the end of each frame. But as best you should turn to persisit entity in render world instead of spawn and clear them every frame.

`ExtractComponentPlugin` now only extracts  "Synchronized" entity to render world.

Co-Authored-By:Trashtalk217 Trashtalk217@gmail.com

re0312 commented:
Added @Trashtalk217  as coauthor ,because  this pr was highly inspired by his work #14252

alice-i-cecile commented:
Marking as Controversial to make sure SME-Rendering signs off on this, but I'm very pleased about this work and want to do whatever we need to get this merged.

re0312 commented:
Following @ecoskey suggestion, Now all unsynchronized entities will be auto cleaned up after the frame ends.

Trashtalk217 reviewed (changes requested):
I wasn't able to finish looking at everything because especially at the end the PR looked like more general refactoring. However, what I've seen looks very good.

My biggest holdup is the ToRenderWorld in spatial_bundle as it seems to be just dead code. And also world_sync.rs needs some extra documentation. Other than that I'm really really impressed by this.

Trashtalk217 commented:
> Following @ecoskey suggestion, Now all unsynchronized entities will be auto cleaned up after the frame ends.

I am not a fan of this. The retained render world is meant to unblock *-as-entities ecs developments. So removing everything *without* a certain component, might match on a component (components-as-entities) and subseqently brick the render world.

I may be misunderstanding something, but this seems to be counterproductive.

re0312 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+pub(crate) fn despawn_fly_entity(world: &mut World) {
+    let mut query = world.query_filtered::<Entity, With<RenderFlyEntity>>();
+
+    // ensure next frame allocation keeps order
```
Some render item(transparent) use Entity as sort_key (e,g bevy_ui), if we cannot guarantee the stability of entity cross frame, it could potentially cause flickering

re0312 commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -1010,16 +1056,18 @@ pub fn prepare_lights(
                 .frusta
                 .get(&entity)
-                .unwrap()
+                .unwrap_or(&empty_frusta)
```
good catch , I fogot to delete it, its unnecessary

re0312 commented on crates/bevy_render/src/spatial_bundle.rs:
```diff
@@ -27,6 +30,8 @@ pub struct SpatialBundle {
     /// The global transform of the entity.
     pub global_transform: GlobalTransform,
+
+    pub _marker: ToRenderWorld,
```
its used for fog_volume or anything utilize spatial_bundle in render world, ToRenderWorld is a explicit component indicate that its entity needs to sync with render world , it maintained by WorldSyncPlugin<ToRenderWorld>

hymm replied:
Feels a bit weird adding this to spatial bundle rather than the specific bundles that need to be extracted. Users often use SpatialBundle when building their own entities, but they might not need to be extracted to the RenderWorld.

re0312 commented on crates/bevy_ui/src/render/render_pass.rs:
```diff
@@ -140,7 +139,7 @@ impl SortedPhaseItem for TransparentUi {
 
     #[inline]
     fn sort(items: &mut [Self]) {
-        items.sort_by_key(SortedPhaseItem::sort_key);
```
we could use a unstable sort here because we can ensure the sort_key is unique ,but looks like I should split it to another pr to minimize the impact of current changes.

alice-i-cecile commented:
Agreed, I think we should not automatically remove things from the render world. It's both a soft-blocker on queries-as-entities and so on, but also generally surprising and implicit.

Trashtalk217 commented on crates/bevy_ui/src/render/mod.rs:
```diff
@@ -205,9 +208,11 @@ pub fn extract_uinode_background_colors(
+    let default_ui_camera = default_ui_camera.get();
     for (
-        entity,
+        _e,
```
What's the point of querying for an entity you don't end up using?

Trashtalk217 commented on crates/bevy_ui/src/render/ui_material_pipeline.rs:
```diff
@@ -327,7 +328,7 @@ impl<P: PhaseItem, M: UiMaterial> RenderCommand<P> for DrawUiMaterialNode<M> {
 
 pub struct ExtractedUiMaterialNode<M: UiMaterial> {
-    pub stack_index: usize,
+    pub stack_index: u32,
```
This change seems unnecessary, but alright

Trashtalk217 commented on crates/bevy_ui/src/render/ui_material_pipeline.rs:
```diff
@@ -664,9 +667,6 @@ pub fn queue_ui_material_nodes<M: UiMaterial>(
         );
-        transparent_phase
-            .items
-            .reserve(extracted_uinodes.uinodes.len());
```
What does this do? Why was it removed?

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+use bevy_hierarchy::DespawnRecursiveExt;
+use bevy_reflect::Reflect;
+
+/// Marker component that indicate its entity needs to be Synchronized to render world
```
```suggestion
/// Marker component that indicates that its entity needs to be Synchronized to the render world
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+        self.0
+    }
+}
+// marker component that its entity needs to be despawned per frame.
```
```suggestion
// marker component that indicates that its entity needs to be despawned at the end of every frame.
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+pub struct RenderFlyEntity;
+
+pub(crate) enum EntityRecord {
+    Added(Entity),
```
```suggestion
    // main
    Added(Entity),
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+    }
+}
+#[derive(Default)]
+pub struct WorldSyncPlugin<B: Bundle> {
```
This would be an amazing place to put some documentation about how the sync plugin/system works. You could use some of the stuff you've already written for the PR.

Trashtalk217 reviewed (changes requested):
Some final sanity checks and another request for more documentation for `world_sync.rs`

ecoskey commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+                        }
+                    }
+                    EntityRecord::Removed(e1, e2) => {
+                        mapper.remove(&e1);
```
Won't this cause entities to be deleted prematurely if a single one of their components is removed? If I understand correctly OnRemove triggers when any of the components in its bundle is removed, right? If this is a problem I'm assuming it hasn't been caught because out internal systems often despawn the whole entity at once.

ecoskey replied:
We might want to have something more fine-grained here.

re0312 replied:
if you haven't added tracking for a component (using WorldSyncPlugin<Component) , the removal of this component will not trigger this event.

ecoskey replied:
That's not the issue, unless I'm misunderstanding. Currently even if you remove a single component it'll delete the render world entity it seems like?

re0312 replied:
The synchronization relationship will  lost only when you delete a specific component . This loss of synchronization is confined to the lifetime scope of that particular component.

For example
e1   [Camera,Camera3D,VisibleEntities,*] 

Because we only added WorldSyncPlugin<Camera>, the synchronization relationship will persist even if Camera3D or VisibleEntities are removed

ecoskey replied:
What's the behavior if two synchronized components are on the same entity, and one is removed? It may not be the "canon" use of `WorldSyncPlugin` but it probably shouldn't be *disallowed*.

ecoskey commented on crates/bevy_render/src/extract_component.rs:
```diff
@@ -209,12 +210,12 @@ impl<T: Asset> ExtractComponent for Handle<T> {
 fn extract_components<C: ExtractComponent>(
```
I think need to give more attention to ExtractComponent since it opens us up to leaving unsynchronized components lying around. Actually this applies to all of our other systems where filters are involved, though I haven't had the chance to go through all of those yet. For example, if an entity matches the ExtractComponent query filter in one frame and the next it doesn't, we're potentially leaving garbage data around.

re0312 replied:
yeah Currently, it appears that ExtractComponentPlugin is only used for extracting components used in Camera View. But it can indeed be used to extract components from other entities as well. Perhaps we should give it a different name to better indicate its usage?

ecoskey commented on crates/bevy_render/src/extract_component.rs:
```diff
@@ -225,13 +226,13 @@ fn extract_components<C: ExtractComponent>(
 fn extract_visible_components<C: ExtractComponent>(
```
Same here. Before, the nuke would handle culling our previously-visible lights but now we don't have that "luxury"

ecoskey commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,126 @@
+/// Marker component that indicate its entity needs to be Synchronized to render world
+#[derive(Component, Clone, Debug, Default, Reflect)]
+#[reflect[Component]]
+pub struct ToRenderWorld;
```
Is adding this the user's responsibility? That seems like it could be a significant footgun, but maybe I've just missed where/if it's automatically added. I'd love if this were just part of ExtractComponentPlugin. And actually, does this have any use outside of being a catch-all to trigger WorldSyncPlugin's observers?

re0312 replied:
No, this is just a manual tagging, and in most cases it is not necessary to use it. If users need to synchronize the entity of a specific component, they should directly use WorldSsyncPlugin<Component>,but it can also be useful in certain situations(`SpatialBundle`)

ecoskey commented on crates/bevy_gizmos/src/lib.rs:
```diff
@@ -432,6 +433,7 @@ fn extract_gizmo_data(
             },
             (*handle).clone_weak(),
             GizmoMeshConfig::from(config),
+            RenderFlyEntity,
```
We discussed this before and it still might be the best option for now, but it feels not great to require users to remember this while extracting.

ecoskey commented on crates/bevy_sprite/src/render/mod.rs:
```diff
@@ -372,7 +373,7 @@ pub fn extract_sprites(
                 slices
                     .extract_sprites(transform, entity, sprite, handle)
-                    .map(|e| (commands.spawn_empty().id(), e)),
+                    .map(|e| (commands.spawn(RenderFlyEntity).id(), e)),
```
Again this seems like a very easy footgun to accidentally run into. Not that I have any better ideas at the moment other than maybe adding the component automatically in ExtractComponentPlugin.

re0312 replied:
yeah, just as you mentioned on Discord, manual tagging can easily to leak entity in some cases.

ecoskey reviewed (changes requested):
Overall LGTM except for some questions about the behavior of WorldSyncPlugin and what I think is potential for mismanaging filtered extraction. I'd also love to see performance numbers for a scene with a large number of extracted entities like lights, additionally with components requiring entity mapping. Currently we don't extract many components at all, so if the sync system has significant overhead I don't think we'd see it in any of the example scenes.

superdump commented:
I want to at least get a chance to look at this before it is merged. And I probably won’t be able to for at least a week.

hymm commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -317,43 +333,71 @@ pub fn extract_lights(
+                if let Some(entity) = mapper.get(e) {
+                    extracted_cascades.insert(*entity, v.clone());
+                } else {
+                    return;
```
should this be a continue? Feels like we should only skip the current entity if for some reason it's missing. Ditto below.

re0312 replied:
In theory, this should not fail. If it fails, the subsequent code will definitely panic.so we should skip this light

hymm commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -317,43 +333,71 @@ pub fn extract_lights(
-            render_visible_entities,
-        ));
-    }
+            // TODO: re-use the memory
```
We could probably query for the existing entity hash maps and reuse them if they exist.

re0312 replied:
yeah , thats the future plan ,if we got a retain render world .Most of the existing extraction logic can directly update components in place instead of reinserting them

hymm reviewed (commented):
I still need to review some of the extraction methods more closely and do a bunch of manually testing.

Could we just change `RenderEntity` to `RenderEntity(Option<Entity)` and use when that component is added or removed to spawn or despawn entities on the render world? Feels like that would be more straightforward than keeping track of all the needed components with  `WorldSyncPlugin`.

re0312 commented:
> Could we just change `RenderEntity` to `RenderEntity(Option<Entity)` and use when that component is added or removed to spawn or despawn entities on the render world? Feels like that would be more straightforward than keeping track of all the needed components with `WorldSyncPlugin`.

In the initial version, I used Option<Entity>, but found that it required manually adding RenderEntity (None) when spawn any entity that needed to be synchronized, and accessing RenderEntity will introduce some boilerplate code in many places. 

```rust
let Some(entity) = render_entity.entity() else{
  return
}
```
Moreover, IMO synchronization behavior should focus on specific component data streams. When a component is removed, its associated synchronization behavior should automatically terminate, rather than persisting until the entity despawns or requiring manual removal of the RenderEntity component.
,
> If I remove all the components off of a point light, but don't despawn it. And then add all the components needed for a spot light to that entity what happens? Does it end up extracted with components from both the point light and the spot light?
> 
In this situation, the behavior of this PR is as expected. Once you remove the pointlight component, the synchronization relationship also ends. When you add a new spotlight component, a new synchronization relationship is established (which is now considered as spotlight).
The possible situation that may cause exceptions is that you add the spotlight component to the pointlight bundle first, and then delete the pointlight component. but we can emit a warning/error when such a situation occurs.

re0312 commented on crates/bevy_pbr/src/cluster/mod.rs:
```diff
@@ -511,7 +512,8 @@ pub(crate) fn clusterable_object_order(
     mut commands: Commands,
-    views: Extract<Query<(Entity, &Clusters, &Camera)>>,
+    views: Extract<Query<(&RenderEntity, &Clusters, &Camera)>>,
+    mapper: Res<MainToRenderEntityMap>,
```
`MainToRenderEntityMap ` is used to manually correct the internal relationships dependencies between some components and other entities.the current implementation is quite ugly, perhaps we can have a dedicated system to correct these.

We use it here because the entity ID within the `Clusters `component is the entity ID of the light in the main world, and we must synchronize it with the ID of the render world

hymm replied:
I'm probably misunderstanding something, but doesn't `query_all_render_entity.get(main_entity)` and `mapper.get(main_entity)` evaluate to the same entity?

re0312 replied:
yeah, you are right ,I misunderstood your meaning before.`MainToRenderEntityMap `is unnecessary

re0312 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,129 @@
+pub struct ToRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug)]
+#[component(storage = "SparseSet")]
```
cargo run --release --example many_lights  --features bevy/trace_tracy 

"extract_lights"
![image](https://github.com/user-attachments/assets/60e4042e-9183-425a-b309-97d8314d464c)
There are approximately 100000 pointlights , with a performance loss of around 10%

But with #14049 ,we could mitigate this performance loss by using `for_each `style iteration.

re0312 commented:
> I'd also love to see performance numbers for a scene with a large number of extracted entities like lights, additionally with components requiring entity mapping.

many_buttons  ~30% loss
![image](https://github.com/user-attachments/assets/0640a387-9e24-41c6-b064-efed8b0e7299)

This example represents the most significant performance loss I've identified. In each frame, it necessitates remapping 80k component dependencies and the spawn and despawn of 80k temporary entities.
The majority of the performance loss from the spawn(~3ms) and despawn(~3ms) operations of ephemeral entity.

alice-i-cecile commented on crates/bevy_pbr/src/lib.rs:
```diff
@@ -319,18 +320,23 @@ impl Plugin for PbrPlugin {
                 FogPlugin,
-                ExtractResourcePlugin::<DefaultOpaqueRendererMethod>::default(),
-                ExtractComponentPlugin::<ShadowFilteringMethod>::default(),
+                (
```
Mild agreement with WrongShoe's style suggestion.

Trashtalk217 reviewed (approved):
All my nits have been resolved for this PR and i think it should be merged. Good work @re0312!

hymm commented:
The 30% perf loss is a bit concerning. Not necessarily for that example, but for games that might want to spawn and despawn things every frame. Where does that perf loss take place? Is it in the render thread when the commands are applied?

StarArawn commented:
> The 30% perf loss is a bit concerning. Not necessarily for that example, but for games that might want to spawn and despawn things every frame. Where does that perf loss take place? Is it in the render thread when the commands are applied?

This is concerning for `bevy_ecs_tilemap` as we treat all tiles as entities. Those don't necessarily get spawned/despawned every frame, but that could be highly game dependent. Another place where this might cause friction entity level UI where for large UI(like an editor) you might be spawning/despawning a large sub-tree of widget/ui entities.

re0312 commented:
>  Where does that perf loss take place? 

The loss from ui_renderer spawning and despawning a large number of temporary entities per frame (~80k).
Mainly in the following two places
depawn_fly_entity(~3ms)  
extract_command_apply&prepare_uinode_command_apply (~3ms)

> Is it in the render thread when the commands are applied?

yeah

>  Not necessarily for that example, but for games that might want to spawn and despawn things every frame

TBH ,I don't think generating and eliminating a large number of entities per frame is the correct behavior. The loss was less apparent before  because we were directly droping the rendered world.With a retained render world, most entities should persist across frames

for some specific games, if #13683 got solved , it should be able to reduce  some of the losses

re0312 commented:
>Another place where this might cause friction entity level UI where for large UI(like an editor) you might be spawning/despawning a large sub-tree of widget/ui entities.

It is worth noting that in this situation, the PR will not have any impact because the `uinode ` is not synchronized, so the spawn/despawn  of UI nodes in the main world will not cause any performance negative effects

However ,if you frequently spwan/despawn a large number of synchronized entities (such as light, camera), the impact of this PR will exceed 30%, and even reach to 100%~150%, this is because you need to spwan/despawn synchronized  entities in two worlds and accompany the trigger of a large number of Observer System

Elabajaba commented:
This is a ~20% performance hit for me when rendering a midsized scene (tested against the synty polygon fantasy kingom demo scene).

re0312 commented:
> This is a ~20% performance hit for me when rendering a midsized scene (tested against the synty polygon fantasy kingom demo scene). We're already really slow here, and this makes one of the slowest and most performance sensitive parts of the engine even slower.

Is there a repository available to test this scene ? There shouldn't be a 20% performance loss once the scene is loaded( or maybe I messed up something).

re0312 commented:
> Is there a repository available to test this scene ? There shouldn't be a 20% performance loss once the scene is loaded( or maybe I messed up something).

I tested the performance  of this pr on [bistro](https://github.com/DGriffin91/bevy_bistro_scene), but I didn't see significant performance loss.

Windows 11 ,13600kf ,Nv4070Ti ,
![image](https://github.com/user-attachments/assets/f4263015-dabb-47ff-984c-8eb78b4e9926)

JMS55 commented:
I believe meshlets probably needs some fixes and performance testing as well. Please reach out to me on discord in #rendering-dev if you have any questions.

https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/src/meshlet/gpu_scene.rs#L40-L143

Olle-Lukowski commented on crates/bevy_pbr/src/cluster/mod.rs:
```diff
@@ -511,7 +512,8 @@ pub(crate) fn clusterable_object_order(
     mut commands: Commands,
-    views: Extract<Query<(Entity, &Clusters, &Camera)>>,
+    views: Extract<Query<(&RenderEntity, &Clusters, &Camera)>>,
+    mapper: Extract<Query<&RenderEntity>>,
```
Can you explain to me what `mapper` is for?

ecoskey replied:
AFAIK it's to remap components with `Entity` fields, changing them from the main world id to the render world id. This is needed since ids are no longer stable across worlds

Olle-Lukowski replied:
Yeah makes sense

Olle-Lukowski commented on crates/bevy_ui/src/render/mod.rs:
```diff
@@ -342,11 +348,24 @@ pub fn extract_uinode_images(
             continue;
         }
 
+        let ui_logical_viewport_size = camera_query
+        .get(camera_entity)
+        .ok()
+        .and_then(bevy_render::camera::Camera::logical_viewport_size)
+        .unwrap_or(Vec2::ZERO)
+        // The logical window resolution returned by `Window` only takes into account the window scale factor and not `UiScale`,
+        // so we have to divide by `UiScale` to get the size of the UI viewport.
+        / ui_scale.0;
```
Is there a reason this was moved?

re0312 replied:
It was obtained by the camera_entity in the main world, so it should be executed before mapping

robtfm commented:
@Elabajaba the perf issues should be fixed apparently, could you retry the synty scene when you have a change?

hymm commented:
Finally getting around to doing some of my own testing.

I'm seeing a 6% loss in many foxes. It seems to be coming from some pretty serious regressions in the perf of the batch_and_prepare_* systems.

![image](https://github.com/user-attachments/assets/02549619-3033-40df-8d79-a121c1ba8b3f)

Still trying to understand why this would be.

re0312 commented:
> Still trying to understand why this would be.

![image](https://github.com/user-attachments/assets/42f5c8b1-7ab2-41fb-afa7-943cd5aca14e)

Looks like a legacy issues, all existing entities will be iterated here,but we should only select view entities

re0312 commented:
#14713 should address the regression mentioned in @hymm  comment. I submitted it as a separate pr since it's unrelated to current.

hymm commented:
> https://github.com/bevyengine/bevy/pull/14713 should address the regression mentioned in @hymm comment. I submitted it as a separate pr since it's unrelated to current.

This pr plus that pr fixes the regression for me in many_foxes. Perf is pretty much identical now.

JMS55 commented:
Meshlets still needs fixing. It maintains hashmaps of entity -> stuff across frames, used for both caching GPU resources, but also accessing the previous frame's resources as it relies on the previous frame's data for culling.

cBournhonesque commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,124 @@
+pub(crate) enum EntityRecord {
+    // main
+    Added(Entity),
+    // render
```
```suggestion
    // When an entity is despawned on the main world, notify the render world so that the corresponding entity can be despawned. This contains the render world entity.
```

cBournhonesque commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,124 @@
+pub struct RenderFlyEntity;
+
+pub(crate) enum EntityRecord {
+    // main
```
```suggestion
    // When an entity is spawned on the main world, notify the render world so that it can spawn a corresponding entity. This contains the main world entity
```

cBournhonesque commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,124 @@
+}
+
+#[derive(Component, Deref, Clone, Debug)]
+pub struct MainEntity(Entity);
```
```suggestion
/// Marker component added on the render world entities to keep track of the corresponding main world entity
pub struct MainEntity(Entity);
```

cBournhonesque commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,124 @@
+pub struct ToRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug, Copy)]
+pub struct RenderEntity(Entity);
```
```suggestion
/// Marker component added on the main world entities that are synced to the Render World in order to keep track of the corresponding render world entity
pub struct RenderEntity(Entity);
```

cBournhonesque commented on crates/bevy_ui/src/render/mod.rs:
```diff
@@ -585,7 +600,7 @@ pub fn extract_uinode_borders(
 
         extracted_uinodes.uinodes.insert(
-            commands.spawn_empty().id(),
+            commands.spawn(RenderFlyEntity).id(),
```
Sorry for the noob question; what are all these `RenderFlyEntities`? They are just temporarily used in the render world for one frame?

re0312 replied:
yeah ,we are currently using many temporary entities for rendering which need to be despawned after frame end.

JMS55 replied:
Imo RenderFlyEntity  -> TemporaryRenderEntity

tychedelia replied:
`TransientEntity` ?

JMS55 replied:
I think it should mention render in the name. Could also go the other way "MarkedForRenderWorldCleanup" or such.

JMS55 replied:
I still want to rename this

tychedelia commented:
It would be nice to add an example that demonstrates writing a custom extract with these changes.

JMS55 commented:
Yeah I'm confused why most extract systems didn't change. I'd expect a _lot_ more code changes from this PR. A clearer migration guide would help me understand a lot more.

ecoskey commented:
What's the intended behavior if an entity has more than one synced component at a time? It seems like removing either one would fully despawn the entity. It would be pretty simple to address this with a count of synced components inside RenderEntity.

alice-i-cecile commented:
For future reference: keeping track of "does this entity have at least one component that requires syncing" is a great fit for first-party trait queries: https://github.com/bevyengine/rfcs/pull/39

re0312 commented:
Ok, Current state of this pr :
 
### Update :
-  Synchronization now focus on entity rather than a component . `SyncRenderWorld` is a marker component that indicates its entity needs to be Synchronized to the render world, this component persist throughout the entity's entire lifecycle. 
This can solve the issue commented by @ecoskey and could work better with  `Required Component ` in the future.
- Fixed the performance regression reported by @Elabajaba .
- Fixed the meshlet issues (still need verification by @JMS55 
- Updated Migration Guide.

### TODOs
- [ ] Add Next Steps
- [ ] Whether to add an example or leave it to future PR ? (TBH I'm not sure how to build an example about this)
- [ ] naming things

NthTensor commented:
The big thing this actually appears to be lacking is documentation. Alice mentioned wanting a broad overview of how extraction works. This is, imo, a blocking requirement. It should probably live in module docs and (contain at-least one easily broken doc-test so we are forced to keep it up to date).

JMS55 commented:
Meshlets seem good now based on how the new extraction system has been described to me. I'll make further improvements to cache resources on view entities directly in a followup PR. 

In terms of getting this merged, I agree we need a lot more docs on this. I don't have time to do another review pass over the code and test all the examples, but that will also be necessary.

ecoskey commented:
I like the move to a single `SyncRenderWorld` component, and the possibility of using required components for this, especially now that that PR's been merged. It definitely simplifies the mental model and keeping track of what component actually marks an entity for syncing.

IMO there's still some things worth ironing out in a followup PR (like a removal/filter check in ExtractComponent), but I'm generally in approval now.

wilk10 commented:
Coming from a user point of view, I have read the PR and superficially followed the discussion on Discord, but I don't know all the nitty gritty. 

I think there's something missing in the documentation about entities that are not synchronized. Are all entities that are not synced (`Without<SyncRenderWorld>`) temporary entities? Or are there three groups: synced, temporary and neither synced nor temporary? I guess the confusion comes from the asymmetry between Sprites being synced and Camera2d not being synced, while 3D Meshes are not synced, but Camera3d is. It's a bit confusing why that is the case. And what about 2D Meshes?

re0312 commented:
>Are all entities that are not synced (`Without<SyncRenderWorld>`) temporary entities?

No , Temporary entities are merely part of entities in  render world . previously, bevy's renderer spawns numerous temporary entities for rendering per frame , which need to be cleared at the end of each frame. IMO, it should  be unnecessary once retain render world landed,
However, to minimize the impact of this PR, I have retained this concept and left it to future pr.

> Or are there three groups: synced, temporary and neither synced nor temporary? 

There are only two groups of entities : synced and unsynced , A synced entity X is one in the main world that has a corresponding ID Y in the render world( X !=Y ).
for a synced main world entity it corresponding render entity ID is stored in the component `RenderEntity`.

> Camera2d not being synced

That's my bad ,Camera2D  is synced but not mentioned in the pr description.

>  while 3D Meshes are not synced, but Camera3d is. It's a bit confusing why that is the case. And what about 2D Meshes?

this is because the current 3D and 2D Mesh instances in render world are all organized within  an EntityHashmap, there are no corresponding render entities at all . so it is unnecessary to sync them.

wilk10 commented on crates/bevy_render/src/world_sync.rs:
````diff
@@ -0,0 +1,252 @@
+/// |      |         |         Render wrold loop                         |
+/// |--------------------------------------------------------------------|
+/// ```
+///
````
Here we could cover the fact that not all entities that are extracted in the render world need to be kept in sync via `SyncRenderWorld` because they are handled differently (without entities) in the render world.

wilk10 commented on examples/3d/fog_volumes.rs:
```diff
@@ -43,7 +44,9 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
             ..default()
-        });
+        })
+        // indicates that this spatial needs to be Synchronized to the render world
```
```suggestion
        // indicates that this fog volume needs to be Synchronized to the render world
```

wilk10 commented:
Thanks for the reply! It's all clear now. I left a comment about specifying it in the documentation as well.

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,252 @@
+
+/// A Plugin that synchronizes entities with specific Components between the main world and render world.
+///
+/// Bevy's renderer is architected independently from main app, It operates in its own separate ECS World. Therefore, the renderer could run in parallel with main app logic, This is called "Pipelined Rendering". See [`PipelinedRenderingPlugin`] for more information.
```
```suggestion
/// Bevy's renderer is architected independently from the main app. It operates in its own separate ECS World, so the renderer could run in parallel with the main app logic. This is called "Pipelined Rendering", see [`PipelinedRenderingPlugin`] for more information.
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,252 @@
+///
+/// Previously, `extract` will copy the related main world entity and its data into the render world , and then render world will clear all render entities at the end of frame to reserve enough entity space to ensure that no main world entity ID has been occupied during next `extract`.
+///
+/// With `* as entities`, we should not clear all entities in render world because some core metadata (e.g. [`Component`], [`Query`]) are also stored in the form of entity.
```
Documentation should describe something as-is, not as a before and after. That's what patch notes are for. These previous two lines can be removed.

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,252 @@
+///
+/// With `* as entities`, we should not clear all entities in render world because some core metadata (e.g. [`Component`], [`Query`]) are also stored in the form of entity.
+///
+/// So we turn to an entity-to-entity mapping strategy to sync between main world entity and render world entity, where each `synchronized` main entity has a [`RenderEntity`] component holds an Entity ID pointer to its unique counterpart entity in the render world.
```
```suggestion
/// [`WorldSyncPlugin`] maintains an entity-to-entity mapping to sync a main world entity and a render world entity. Each synchronized main entity has a [`RenderEntity`] component that holds an Entity pointer to its unique counterpart entity in the render world.
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,252 @@
+///
+/// So we turn to an entity-to-entity mapping strategy to sync between main world entity and render world entity, where each `synchronized` main entity has a [`RenderEntity`] component holds an Entity ID pointer to its unique counterpart entity in the render world.
+///
+/// A example for `synchronized` main entity 1v1 and 18v1
```
```suggestion
/// An example for synchronized main entities 1v1 and 18v1
```

Trashtalk217 commented on crates/bevy_render/src/world_sync.rs:
````diff
@@ -0,0 +1,252 @@
+///
+/// ```
+///
+/// To establish a "Synchronous Relationship", you can add a [`SyncRenderWorld`] component to an entity. Once a `synchronized` main entity is despawned ,its corresponding render entity will be automatically purged in the next `Sync`
````
```suggestion
/// To establish a "Synchronous Relationship", you can add a [`SyncRenderWorld`] component to an entity. Once a synchronized main entity is despawned, its corresponding render entity will be automatically purged in the next `Sync`
```

Trashtalk217 replied:
This should probably have some example code of how people can sync their own components using this plugin.

Trashtalk217 reviewed (commented):
I've checked the documentation. It could use some work, but it's getting there.

bushrat011899 commented on crates/bevy_core_pipeline/src/auto_exposure/buffers.rs:
```diff
@@ -27,13 +28,13 @@ pub(super) struct ExtractedStateBuffers {
         changed: changed
             .iter()
-            .map(|(entity, settings)| (entity, settings.clone()))
+            .map(|(entity, settings)| (entity.id(), settings.clone()))
```
There's quite a lot of noise in this PR due to the need to insert these `RenderEntity::id` calls. This might be a controversial suggestion, but we _can_ implement `WorldQuery` for `RenderEntity` so it still behaves like a `Component`, but also returns the inner `Entity` when queried directly.

<details>
  <summary>unsafe impl WorldQuery for RenderEntity { ... }</summary>
  
  ```rust
/// SAFETY:
/// This implementation delegates to the existing implementation for &RenderEntity
unsafe impl WorldQuery for RenderEntity {
    type Item<'w> = Entity;
    type Fetch<'w> = ReadFetch<'w, RenderEntity>;
    type State = ComponentId;

    fn shrink<'wlong: 'wshort, 'wshort>(item: Entity) -> Entity {
        item
    }

    fn shrink_fetch<'wlong: 'wshort, 'wshort>(fetch: Self::Fetch<'wlong>) -> Self::Fetch<'wshort> {
        <&Self as WorldQuery>::shrink_fetch(fetch)
    }

    #[inline]
    unsafe fn init_fetch<'w>(
        world: UnsafeWorldCell<'w>,
        component_id: &ComponentId,
        last_run: Tick,
        this_run: Tick,
    ) -> ReadFetch<'w, RenderEntity> {
        // SAFETY: This implementation delegates to the existing implementation for &RenderEntity
        unsafe {
            <&Self as WorldQuery>::init_fetch(world, component_id, last_run, this_run)
        }
    }

    const IS_DENSE: bool = <&Self as WorldQuery>::IS_DENSE;

    #[inline]
    unsafe fn set_archetype<'w>(
        fetch: &mut ReadFetch<'w, RenderEntity>,
        component_id: &ComponentId,
        archetype: &'w Archetype,
        table: &'w Table,
    ) {
        // SAFETY: This implementation delegates to the existing implementation for &RenderEntity
        unsafe {
            <&Self as WorldQuery>::set_archetype(fetch, component_id, archetype, table)
        }
    }

    #[inline]
    unsafe fn set_table<'w>(
        fetch: &mut ReadFetch<'w, RenderEntity>,
        component_id: &ComponentId,
        table: &'w Table,
    ) {
        // SAFETY: This implementation delegates to the existing implementation for &RenderEntity
        unsafe {
            <&Self as WorldQuery>::set_table(fetch, component_id, table)
        }
    }

    #[inline(always)]
    unsafe fn fetch<'w>(
        fetch: &mut Self::Fetch<'w>,
        entity: Entity,
        table_row: TableRow,
    ) -> Self::Item<'w> {
        // SAFETY: This implementation delegates to the existing implementation for &RenderEntity
        unsafe {
            <&Self as WorldQuery>::fetch(fetch, entity, table_row).id()
        }
    }

    fn update_component_access(
        component_id: &ComponentId,
        access: &mut FilteredAccess<ComponentId>,
    ) {
        <&Self as WorldQuery>::update_component_access(component_id, access)
    }

    fn init_state(world: &mut World) -> ComponentId {
        <&Self as WorldQuery>::init_state(world)
    }

    fn get_state(components: &Components) -> Option<Self::State> {
        <&Self as WorldQuery>::get_state(components)
    }

    fn matches_component_set(
        state: &ComponentId,
        set_contains_id: &impl Fn(ComponentId) -> bool,
    ) -> bool {
        <&Self as WorldQuery>::matches_component_set(state, set_contains_id)
    }
}

/// SAFETY: `Self` is the same as `Self::ReadOnly`
unsafe impl QueryData for RenderEntity {
    type ReadOnly = Self;
}

/// SAFETY: access is read only
unsafe impl ReadOnlyQueryData for RenderEntity {}
  ```
  
</details>

This would then allow for render systems to query for `RenderEntity` and directly receive the inner `Entity`, reducing the noise in this PR. Crucially, this doesn't conflict with the existing `derive(Component)`, so you can still query for `&RenderEntity`, the mutable version, use it in filters, etc.

bushrat011899 reviewed (commented):
I'm not familiar enough with the rendering side of Bevy to feel comfortable approving this PR, but it does look good to my novice eyes. I don't see anything that stands out as particularly wrong (that hasn't already been pointed in e.g. documentation).

I _do_ have a suggestion around implementing `WorldQuery` on `RenderEntity` directly, which reduces the noise in this PR quite a bit and gives the rendering people some slightly nicer ergonomics. I've tested this implementation and it works as expected, allowing `RenderEntity` to be a drop-in replacement for `Entity`.

alice-i-cecile commented:
I think I like the WorldQuery impl 🤔

NthTensor commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// |      |         |          Main world loop                          |
+/// | sync | extract |---------------------------------------------------|
+/// |      |         |         Render world loop                         |
+//are/ |--------------------------------------------------------------------|
```
this looks like a typo

alice-i-cecile replied:
```suggestion
/// |--------------------------------------------------------------------|
```

NthTensor reviewed (commented):
I would say this generally meets my minimum requirements for docs.

wilk10 commented on crates/bevy_render/src/world_sync.rs:
````diff
@@ -0,0 +1,261 @@
+///
+/// ```
+///
+/// Note that not this effectively establishes a link between the main world entity and the render world entity.
````
```suggestion
/// Note that this effectively establishes a link between the main world entity and the render world entity.
```

wilk10 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// Note that not this effectively establishes a link between the main world entity and the render world entity.
+/// Not every entity needs to be synchronized, however, only entities with [`SyncRenderWorld`] are synced.
+/// Adding [`SyncRenderWorld`] to a main world component will establish such a link.
+/// Once a synchronized main entity is despawned, it's corresponding render entity will be automatically
```
```suggestion
/// Once a synchronized main entity is despawned, its corresponding render entity will be automatically
```

wilk10 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// The sync step does not handle the transfer of component data between worlds,
+/// since it's often not necessary to transfer over all the components of a main world entity.
+/// The render world probably cares about a `Position` component, but not a `Velocity` component.
+/// The extraction happens in it's own step, independently from synchronization.
```
```suggestion
/// The extraction happens in its own step, independently from synchronization.
```

JMS55 replied:
```suggestion
/// The extraction happens in it's own step, independently from, and after synchronization.
```

wilk10 reviewed (approved):
LGTM! I mainly focused on the documentation and commented on a few tiny doc adjustments. 
Thanks for previously including the part about assets not being synced, it's very clear now.

JMS55 commented on crates/bevy_core_pipeline/src/core_2d/camera_2d.rs:
```diff
@@ -37,6 +38,8 @@ pub struct Camera2dBundle {
     pub deband_dither: DebandDither,
     pub main_texture_usages: CameraMainTextureUsages,
     pub msaa: Msaa,
+    /// Marker component that indicates that its entity needs to be Synchronized to the render world
```
For here and all other comments, please do not capitalize "Synchronized" in the middle of the sentence.

JMS55 commented on crates/bevy_core_pipeline/src/taa/mod.rs:
```diff
@@ -346,17 +347,22 @@ impl SpecializedRenderPipeline for TaaPipeline {
     for (entity, camera, camera_projection, mut taa_settings) in
         cameras_3d.iter_mut(&mut main_world)
     {
+        let entity = entity.id();
```
Tiny nit: entity is only used once here, I'd rather not create a new variable, and just use entity.id() directly in commands.get_or_spawn()

JMS55 commented on crates/bevy_core_pipeline/src/taa/mod.rs:
```diff
@@ -346,17 +347,22 @@ impl SpecializedRenderPipeline for TaaPipeline {
+        let entity = entity.id();
         let has_perspective_projection = matches!(camera_projection, Projection::Perspective(_));
         if camera.is_active && has_perspective_projection {
             commands.get_or_spawn(entity).insert(taa_settings.clone());
```
Lets see if I understand this correctly: this could be just get_mut() here instead of get_or_spawn(), as the entity should already exist as camera bundles have SyncRenderWorld, correct?

(That said I do think we should leave it as get_or_spawn() in case users aren't using the bundle, this comment is just to see if I understand this PR correctly or not)

JMS55 commented on crates/bevy_pbr/src/cluster/mod.rs:
```diff
@@ -531,13 +533,15 @@ pub fn extract_clusters(
-                data.push(ExtractedClusterableObjectElement::ClusterableObjectEntity(
-                    *clusterable_entity,
-                ));
+                if let Ok(entity) = mapper.get(*clusterable_entity) {
```
Shouldn't we modify the clusterizer system to store RenderEntity on Clusters directly, instead of using another query here?

JMS55 commented on crates/bevy_pbr/src/lib.rs:
```diff
@@ -420,6 +420,9 @@ impl Plugin for PbrPlugin {
             )
             .init_resource::<LightMeta>();
 
+        render_app.world_mut().observe(add_light_view_entities);
```
What are these two lines doing here ?

JMS55 commented on crates/bevy_pbr/src/light_probe/mod.rs:
```diff
@@ -433,14 +436,15 @@ fn gather_light_probes<C>(
         // Gather up the light probes in the list.
         render_view_light_probes.maybe_gather_light_probes(&view_reflection_probes);
 
+        let entity = view_entity.id();
```
I'll stop mentioning it now as it's a minor point and once we move to a WorldQuery impl it should go away anyways, but I'd personally prefer to not make extra variables like this for accessing entity.id().

JMS55 commented on crates/bevy_pbr/src/meshlet/instance_manager.rs:
```diff
@@ -178,21 +178,21 @@ pub fn extract_meshlet_mesh_entities(
 
     // Reset per-frame data
-    instance_manager.reset(entities);
+    instance_manager.reset(render_entities);
```
(Not a review comment): Lots of room to heavily improve the design of the meshlet instance and resource managers once this PR gets merged :)

JMS55 commented on crates/bevy_render/src/camera/camera.rs:
```diff
@@ -935,7 +935,8 @@ pub fn extract_cameras(
             }
 
-            let mut commands = commands.get_or_spawn(entity).insert((
+            let mut commands = commands.entity(render_entity.id());
```
Why all the changes to formatting here? Correct me if I'm wrong, but I think there's no real code changes in this file besides the one entity -> render_entity.id(), yes?

Trashtalk217 replied:
Yes, that's correct

Trashtalk217 replied:
Yes, that's the only substantive change

JMS55 commented on crates/bevy_render/src/extract_component.rs:
```diff
@@ -205,33 +206,33 @@ impl<T: Asset> ExtractComponent for Handle<T> {
 }
 
-/// This system extracts all components of the corresponding [`ExtractComponent`] type.
+/// This system extracts all synced entities components of the corresponding [`ExtractComponent`] type.
```
This description does not make sense to me / is unclear. 

Correct me if I'm wrong, but this system/plugin does basically exactly what it did before this PR, _except_ the only new thing is that there's a new constraint: any entity with the component C, must also have SyncRenderWorld on it as well? @alice-i-cecile or ECS SMEs: Can we require that `C` here must also require SyncRenderWorld in the trait bounds? Is `C: Require<SyncRenderWorld>` a thing?

As for fixing the description, assuming I understand everything correctly, I would say "This system extracts all components of the corresponding [`ExtractComponent`] type, for entities that are synced via SyncRenderWorld", or something like that.

JMS55 commented on crates/bevy_render/src/extract_component.rs:
```diff
@@ -205,33 +206,33 @@ impl<T: Asset> ExtractComponent for Handle<T> {
 }
 
-/// This system extracts all visible components of the corresponding [`ExtractComponent`] type.
+/// This system extracts all synced visible entities components of the corresponding [`ExtractComponent`] type.
```
Same here as above about the description.

JMS55 commented on crates/bevy_render/src/lib.rs:
```diff
@@ -473,35 +479,15 @@ unsafe fn initialize_render_app(app: &mut App) {
                 )
                     .in_set(RenderSet::Render),
-                World::clear_entities.in_set(RenderSet::PostCleanup),
+                despawn_temporary_render_entity.in_set(RenderSet::PostCleanup),
```
```suggestion
                despawn_temporary_render_entities.in_set(RenderSet::PostCleanup),
```
Need to rename the system definition too ofc

JMS55 commented on crates/bevy_render/src/pipelined_rendering.rs:
````diff
@@ -83,13 +83,15 @@ impl Drop for RenderAppChannels {
+/// |---------------------------------------------------------------------------|
 /// ```
 ///
+/// - `sync` is the step where the entity-entity mapping between the main and render world is updated.
+///     This is run on the main app's thread, for more information checkout [`WorldSyncPlugin`].
````
```suggestion
/// - `sync` is the step where the entity-entity mapping between the main and render world is updated. For more information checkout [`WorldSyncPlugin`].
///     This is run on the main app's thread, .
```
Or something to this effect. Just slightly group the sentences better (hard to figure out what it will look like on github, but I hope you get my point)

JMS55 commented on crates/bevy_sprite/src/mesh2d/mesh.rs:
```diff
@@ -221,8 +221,6 @@ pub struct RenderMesh2dInstances(EntityHashMap<RenderMesh2dInstance>);
 pub struct Mesh2d;
 
 pub fn extract_mesh2d(
```
What's with these changes? Unclear why at first glance what the original code was doing.

JMS55 commented on crates/bevy_ui/src/render/ui_material_pipeline.rs:
```diff
@@ -374,6 +376,7 @@ pub fn extract_ui_material_nodes<M: UiMaterial>(
     >,
     windows: Extract<Query<&Window, With<PrimaryWindow>>>,
     ui_scale: Extract<Res<UiScale>>,
+    mapping: Extract<Query<&RenderEntity>>,
```
```suggestion
    render_entity_lookup: Extract<Query<&RenderEntity>>,
```

JMS55 commented on examples/3d/fog_volumes.rs:
```diff
@@ -43,7 +44,9 @@ fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
-        });
+        })
+        // indicates that this fog volume needs to be Synchronized to the render world
+        .insert(SyncRenderWorld);
```
Ouch this feels like a big footgun. I _really_ don't like that users have to be aware to add SyncRenderWorld for spatial entities that are parents of renderable-entities.

Nothing's immediately appearing in my brain as a solution, but I'd really like to brainstorm ways to avoid this.

cart replied:
I suspect that this could be switched to a FogVolumeBundle, and that FogVolumeBundle should have SyncRenderWorld? If not, I do agree that this is a pretty major UX issu.

JMS55 commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -176,6 +178,7 @@ pub fn extract_lights(
     global_point_lights: Extract<Res<GlobalVisibleClusterableObjects>>,
     point_lights: Extract<
         Query<(
+            &RenderEntity,
```
I don't have the brainpower to review this file atm. Requesting that one of the other reviewers please carefully review this one as I did not.

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+use bevy_reflect::Reflect;
+use bevy_utils::tracing::warn;
+
+/// A Plugin that synchronizes entities with specific Components between the main world and the render world.
```
```suggestion
/// A Plugin that synchronizes entities with [`SyncRenderWorld`] between the main world and the render world.
```

It's no longer true that syncing is per-component, right?

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// A Plugin that synchronizes entities with specific Components between the main world and the render world.
+///
+/// Bevy's renderer is architected independently from the main app.
+/// It operates in its own separate ECS World, so the renderer could run in parallel with the main world logic.
```
```suggestion
/// It operates in its own separate ECS [`World`], so the renderer logic can run in parallel with the main world logic.
```

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// This is called "Pipelined Rendering", see [`PipelinedRenderingPlugin`] for more information.
+///
+/// [`WorldSyncPlugin`] is the first thing that runs every frame and it maintains an entity-to-entity mapping
+/// between the main world and the render world.
```
```suggestion
/// between the main world and the render world, by spawning new entities (without any components) in the render world to match newly spawned entities in the main world, and despawning entities in the render world to match newly despawned entities in the main world.
```

Could also omit these changes as I see that you mention it more later on in the docs, up to you. But I would like to keep the specific mention that when we spawn new entities, they do not have any components.

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+///
+/// [`WorldSyncPlugin`] is the first thing that runs every frame and it maintains an entity-to-entity mapping
+/// between the main world and the render world.
+/// This is necessary for extraction, which copies over component data from the main to the render world.
```
```suggestion
/// This is necessary preparation for extraction (`[ExtractSchedule]`), which copies over component data from the main to the render world for these entities.
```

JMS55 commented on crates/bevy_render/src/world_sync.rs:
````diff
@@ -0,0 +1,261 @@
+/// ```
+///
+/// Note that not this effectively establishes a link between the main world entity and the render world entity.
+/// Not every entity needs to be synchronized, however, only entities with [`SyncRenderWorld`] are synced.
````
```suggestion
/// Not every entity needs to be synchronized, however; only entities with the [`SyncRenderWorld`] component are synced.
```

Or semicolon, or dash, idk english is hard.

Actually maybe just delete this whole sentence, as I asked to mention it in an earlier sentence/comment.

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// Once a synchronized main entity is despawned, it's corresponding render entity will be automatically
+/// despawned in the next `sync`.
+///
+/// The sync step does not handle the transfer of component data between worlds,
```
```suggestion
/// The sync step does not copy any component data between worlds,
```

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+/// The render world probably cares about a `Position` component, but not a `Velocity` component.
+/// The extraction happens in it's own step, independently from synchronization.
+///
+/// Moreover, [`WorldSyncPlugin`] only synchronizes *entities*, stuff like mesh and texture data is handled
+/// differently, as those assets are extracted to render world resources and not entities.
```
```suggestion
/// Moreover, [`WorldSyncPlugin`] only synchronizes *entities*. [`RenderAsset`]s like meshes and textures are handled differently.
```

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+        app.init_resource::<PendingSyncEntity>();
+        app.observe(
+            |trigger: Trigger<OnAdd, SyncRenderWorld>, mut pending: ResMut<PendingSyncEntity>| {
+                pending.push(EntityRecord::Added(trigger.entity()));
```
Do we ever shrink the size of `pending`?

Trashtalk217 replied:
Yes, with `for record in pending.drain(..) {` every frame.

JMS55 replied:
Sorry not what I meant I should've been more specific; drain just empties the contents, it doesn't resize the _capacity_ (i.e. allocated memory).

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+        );
+    }
+}
+/// Marker component that indicates that its entity needs to be Synchronized to the render world
```
```suggestion
/// Marker component that indicates that its entity needs to be synchronized to the render world.
```

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+pub struct SyncRenderWorld;
+
+#[derive(Component, Deref, Clone, Debug, Copy)]
+/// Marker component added on the main world entities that are synced to the Render World in order to keep track of the corresponding render world entity
```
Doc comments should go above the derive, not below.

Also do we want to make a note to use required components in the future to avoid the archetype move on insert, or?

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+    }
+}
+
+// marker component that indicates that its entity needs to be despawned at the end of every frame.
```
```suggestion
/// Marker component that indicates that its entity needs to be despawned at the end of the frame.
///
/// ...[1]
```

[1]: I'd like if you could go into what the use case would be here, i.e. spawning entities *in the render world* that are only needed for a single frame.

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+pub struct TemporaryRenderEntity;
+
+pub(crate) enum EntityRecord {
+    // When an entity is spawned on the main world, notify the render world so that it can spawn a corresponding entity. This contains the main world entity
```
```suggestion
    /// When an entity is spawned on the main world, notify the render world so that it can spawn a corresponding entity. This contains the main world entity.
```

These also need to be doc comments yeah? Triple slash, not double. Also missing a period.

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+                    if let Some(mut entity) = world.get_entity_mut(e) {
+                        match entity.entry::<RenderEntity>() {
+                            bevy_ecs::world::Entry::Occupied(_) => {
+                                warn!("Attempting to synchronize an entity that has already been synchronized!");
```
warn -> error?

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+}
+
+pub(crate) fn entity_sync_system(main_world: &mut World, render_world: &mut World) {
+    main_world.resource_scope(|world, mut pending: Mut<PendingSyncEntity>| {
```
This function could use some splitting up into closures or smaller functions to reduce the nesting 😬

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+                }
+                EntityRecord::Removed(e) => {
+                    if let Some(ec) = render_world.get_entity_mut(e) {
+                        ec.despawn_recursive();
```
Is recursive despawn what we want here?

JMS55 commented on crates/bevy_render/src/world_sync.rs:
```diff
@@ -0,0 +1,261 @@
+
+    local.extend(query.iter());
+
+    // ensure next frame allocation keeps order
```
```suggestion
    // Ensure next frame allocation keeps order
```

Also, can you expand the comment a bit more? What does this mean, and why are we doing it?

JMS55 commented on crates/bevy_ui/src/render/mod.rs:
I'm also going to pass on this file for now, but we'll need to review it at some point soon.

JMS55 reviewed (changes requested):
I left a bunch of comments, but awesome work overall! This is much clearer and more intuitive than the previous version, and I'm happy with how this looks and the level of docs.

WorldQuery impl would be nice, but we can leave it to a followup (open an issue please!).

My only major concern is needing SyncRenderWorld on intermediate spatial entities - this seems like a huge footgun. Not sure how to solve it...

I'd also vote in favor of renaming SyncRenderWorld to SyncWithRenderWorld or SyncToRenderWorld to be even more explicit, but I don't feel too strongly on it, the name is acceptable as-is.

ecoskey commented:
One last nit from me: while it doesn't affect any of the in-engine uses of `ExtractComponentPlugin` afaik, deletions of components in the main world won't propagate to their render world counterparts. I'd like to at least see a notice that `ExtractComponentPlugin` has this caveat, and recommend to only use it with transient entities or those whose components are not expected to be removed (aside from when the entity is despawned)

JMS55 commented:
Something else I forgot to mention, I'd like to mention in the code somewhere that we _have_ to make sure that RenderEntity IDs are consistent across frames, i.e. if entity A has render entity B in one frame, then that same mapping will exist in all subsequent frames. A test for it would be awesome as well if you feel up to it.
